flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/AbstractCodeGeneratorCastRule.java

```diff
@@ -102,12 +105,32 @@ protected AbstractCodeGeneratorCastRule(CastRulePredicate predicate) {
         final String functionSignature =
                 "@Override public Object cast(Object _myInputObj) throws "
                         + className(TableException.class);
-        final String inputVarDecl =
-                inputTypeTerm + " " + inputTerm + " = (" + inputTypeTerm + ") _myInputObj;\n";
-        final String inputIsNullVarDecl =
-                "boolean " + inputIsNullTerm + " = _myInputObj == null;\n";
 
-        final String returnStmt = "return " + codeBlock.getReturnTerm() + ";\n";
+        // Write the function body
+        final CastRuleUtils.CodeWriter bodyWriter = new CastRuleUtils.CodeWriter();
+        bodyWriter.declStmt(inputTypeTerm, inputTerm, cast(inputTypeTerm, "_myInputObj"));
+        bodyWriter.declStmt("boolean", inputIsNullTerm, "_myInputObj == null");
+        ctx.variableDeclarationStatements.forEach(decl -> bodyWriter.appendBlock(decl + "\n"));
+
+        if (this.canFail()) {
+            bodyWriter.tryCatchStmt(
+                    tryWriter ->
+                            tryWriter.append(codeBlock).stmt("return " + codeBlock.getReturnTerm()),
+                    Throwable.class,
+                    (exceptionTerm, catchWriter) ->
+                            catchWriter.throwStmt(
+                                    constructorCall(
```

Comment:
Maybe it makes sense to also provide the exception instantiation also as a utility method.

Reply:
You mean something like `newTableException(args...)` which translates to `constructorCall(TableException.class, args...)`?

Reply:
Yep, or even a more generic `newException(Class<?> exceptionClass, String msg)`

Reply:
or even do everything in the `throwStmt`? It's just a suggestion to maybe have even more "clean" code here.

Reply:
I don't know, TBH I don't like much `newException(Class<?> exceptionClass, String msg)` because then we need the version of the method that accepts the cause term, making the signature a bit confusing because both are strings: `newException(Class<?> exceptionClass, String msg, String causeTerm)`. Adding the logic to `throwStmt` is also inconsistent with the rest of `CodeWriter`, as `throw` accepts any exception, not just a constructor invocation expression...

What if we revisit this topic once we'll need in some rule to throw an exception in the cast logic? This `AbstractCodeGeneratorCastRule` is a bit of a special case for me because it just glues the code generated rules with the `CastExecutor`, but it doesn't really have cast logic, so I'm not sure whether we'll ever need again this try catch and throw new methods and how we want to use them

Reply:
Yep yep, not insisting at all here, was just a suggestion if it would help.